src/main/java/org/checkerframework/specimin/JavaParserUtil.java

```diff
@@ -1365,10 +1406,269 @@ public static String getInitializerRHS(String variableType) {
           }
         }
       }
-      // The handling of methods in anonymous classes seems to work right now. If an issue
+      // The current handling of methods in anonymous classes seems to work for now. If an issue
       // arises in the future, add it here.
       // TODO: add method finding based on name and parameter types
     }
     return null;
   }
+
+  /**
+   * Returns the resolved declaration of the expression if expression.resolve() throws an
+   * UnsupportedOperationException due to its qualifying expression being a constraint type. Returns
+   * null if any solving fails.
+   *
+   * <p>An example: resolving {@code foo.method()}, where {@code foo} is of type {@code ? extends
+   * T}. Resolving the method call will fail because T is a type parameter, not a declaration. This
+   * could occur in a lambda, where foo is a lambda parameter that is of type {@code ? extends T},
+   * but this specific lambda has a type argument for T (like Foo). In this case, we'll want to find
+   * the corresponding method declaration from {@code Foo}. A similar case can be found in
+   * LambdaBodyStaticUnsolved2Test.
+   *
+   * <p>This method works under the assumption that the type argument type is solvable; if not, this
+   * will return null.
+   *
+   * @param expression The expression to find the declaration of (method, field)
+   * @return The resolved object, if it exists; null otherwise
+   */
+  public static @Nullable Object tryFindCorrespondingDeclarationForConstraintQualifiedExpression(
```

Comment:
I didn't read this that closely and will need to review this again when I have a bit more mental energy available. I don't see anything obviously wrong for now, though.